src/http/server/handlers/static_file_handler.cr

```diff
@@ -99,15 +101,81 @@ class HTTP::StaticFileHandler
         end
       end
 
-      context.response.content_length = file_info.size
       File.open(file_path) do |file|
-        IO.copy(file, context.response)
+        if range_header = context.request.headers["Range"]?
+          range_header = range_header.lchop?("bytes=") || return context.response.respond_with_status :range_not_satisfiable
+          ranges = parse_ranges(range_header, file_info.size) || return context.response.respond_with_status :bad_request
+
+          # If any of the ranges start beyond the end of the file, we return an
+          # HTTP 416 Range Not Satisfiable.
+          # See https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests#partial_request_responses
+          if ranges.any? { |range| range.begin > file_info.size }
+            context.response.headers["Content-Range"] = "bytes */#{file_info.size}"
+            context.response.status = :range_not_satisfiable
+            context.response.close
+            return
+          end
+
+          context.response.status = :partial_content
+
+          if ranges.size == 1
+            range = ranges.first
+            file.seek range.begin
+            context.response.headers["Content-Range"] = "bytes #{range.begin}-#{range.end}/#{file_info.size}"
+            IO.copy file, context.response, range.size
+          else
+            MIME::Multipart.build(context.response) do |builder|
+              content_type = context.response.headers["Content-Type"]?
+              context.response.headers["Content-Type"] = builder.content_type("byterange")
+
+              ranges.each do |range|
+                file.seek range.begin
+                headers = HTTP::Headers{
+                  "Content-Range"  => "bytes #{range.begin}-#{range.end}/#{file_info.size}",
+                  "Content-Length" => range.size.to_s,
+                }
+                headers["Content-Type"] = content_type if content_type
+                chunk_io = IO::Sized.new(file, range.size)
+                builder.body_part headers, chunk_io
+              end
+            end
+          end
+        else
+          context.response.status = :ok
+          context.response.content_length = file_info.size
+          IO.copy(file, context.response)
+        end
       end
     else # Not a normal file (FIFO/device/socket)
       call_next(context)
     end
   end
 
+  # TODO: Optimize without lots of intermediary strings
+  private def parse_ranges(header, file_size)
+    ranges = [] of Range(Int64, Int64)
+    header.split(",") do |range|
+      start_string, dash, finish_string = range.lchop(' ').partition("-")
+      return if dash.empty?
+      start = start_string.to_i64?
+      return if start.nil? && !start_string.empty?
+      if finish_string.empty?
+        finish = file_size
+      else
+        finish = finish_string.to_i64? || return
+      end
+      unless start
+        start = file_size - finish
+        finish = file_size
```

Comment:
From [here](https://www.rfc-editor.org/rfc/rfc9110.html#section-14.1.2-7):

> A client can refer to the last N bytes (N > 0) of the selected representation using a [suffix-range](https://www.rfc-editor.org/rfc/rfc9110.html#rule.suffix-range). If the selected representation is shorter than the specified [suffix-length](https://www.rfc-editor.org/rfc/rfc9110.html#rule.suffix-range), the entire representation is used.

So a better implementation is:

```suggestion
        return unless finish > 0
        start = {file_size - finish, 0_i64}.max
        finish = file_size - 1
```

Reply:
Good catch!

I don't think the `return` is necessary, though. If `finish <= 0`, the range would end up decreasing and thus invalid.